src/runtime/hexagon_dma_device_interface.cpp

```diff
@@ -0,0 +1,441 @@
+        region_tcm_desc_size = dma_get_descriptor_size(ae_fmt_id,ncomponents,nfolds);
+
+        //#########################################################
+        //Now allocate the L2 Cache
+        //########################################################
+        //The maximum TCM Buf Size
+        qurt_size_t tcm_buf_size;
+        if(halide_hexagon_dmart_get_fold_size(user_context,(uintptr_t)buf->host,&tcm_buf_size)) {
+            error(NULL) << "Function failed to get Fold Buffer Size \n";
+            return E_ERR;
+        }
+        /* Allocating in multiples of 4K */
+        qurt_size_t region_tcm_size = align(tcm_buf_size, 0x1000);
+
+        // It is a good idea to check that this size is not too large, as while we are still in DDR, when locked to the TCM
+        // large sizes will become problematic.
+        qurt_size_t region_tcm_limit = 0x40000; // Limit is set to 256k
+        if (region_tcm_size > region_tcm_limit) {
+            error(NULL) << "The required TCM region for this ROI" << region_tcm_size << \
+                           "exceeds the set limit of " << region_tcm_limit;
+            error(NULL) << "The ROI must be lowered or the allowed region made larger.\n" ;
+            return E_ERR;
+        }
+
+        nRet = dma_get_mem_pool_id(&pool_tcm);
+        if(nRet != QURT_EOK) {
+            error(NULL) << "Failed to attach the TCM memory region. The error code is:" << nRet << "\n" ;
+            return E_ERR;
+        }
+
+        tcm_buf_vaddr = dma_allocate_cache(pool_tcm, region_tcm_size, &region_tcm);
+        tcm_desc_vaddr = dma_allocate_cache(pool_tcm, region_tcm_desc_size, &region_tcm_desc);
+
+        // Lock the TCM region. This maps the region mapped as TCM to the actual TCM.
+        // This is done to ensure the region is not invalidated during DMA processing.
+        // This can also be done through the firmware (using the nDmaWrapper_LockCache function).
+        nRet = dma_lock_cache(tcm_buf_vaddr, region_tcm_size);
+        if(nRet != QURT_EOK) {
+            error(NULL) << "QURT TCM lock failed due to QURT_EALIGN ERROR  \
+                           misaligned u32Size = " << region_tcm_size << "\n";
+            return E_ERR;
+        }
+
+        // Lock the descriptor region as well.
+        nRet  = dma_lock_cache(tcm_desc_vaddr, region_tcm_desc_size);
+        if(nRet != QURT_EOK) {
+            error(NULL) << "QURT TCM lock failed due to QURT_EALIGN ERROR  \
+                           misaligned u32Size = " << region_tcm_desc_size << "\n" ;
+            return E_ERR;
+        }
+
+        if(halide_hexagon_dmart_set_fold_storage(user_context, tcm_buf_vaddr, region_tcm, \
+            tcm_buf_size, tcm_desc_vaddr, region_tcm_desc, region_tcm_desc_size, &fold_idx)) {
+            error(NULL) << "Function failed to Set Fold Storage to DMA context \n";
+            return E_ERR;
+        }
+
+        /*Assign the Allocated Fold Storage to Device Memory*/
```

Comment:
There's a mix of `//` style comments and `/* */` style comments. I think Halide convention is to use `//` everywhere, except for doxygen comments. It definitely should be consistent!